feat: TNL-10136 tease course enrollment from student model

bszabo opened the pull request:
Reapply changes developed in bszabo/TNL-10136-student-course-enrollment branch to current master

The intent here is to resolve dependency version conflicts only, relative to PR https://github.com/openedx/edx-platform/pull/31229

Commit:
feat: TNL-10136 tease course enrollment from student model

Reapply changes developed in bszabo/TNL-10136-student-course-enrollment branch to current master

Commit:
feat: TNL-10136 course_enrollment.py hygiene


inclusion in __init__
docstring

bszabo committed:
feat: TNL-10136 student.py -> user.py

Commit:
feat: TNL-10136 fix lint errors


And import cross-references
And wrap-around lines

Commit:
feat: TNL-10136 Propagate changes from last PR

Commit:
feat: TNL-10136 fix build errors

blank lines in __init__
private method that needed to be public
importlib and importlib-resources dependencies

bszabo commented:
@connorhaugh This PR is now ready for your review. It should have no substantive changes from [the previous one you reviewed](https://github.com/openedx/edx-platform/pull/31229). That one had been approved but ran into dependency conflicts with master. The incremental changes here are to resolve those dependency changes.

connorhaugh commented:
Ok @bszabo I will read through again to double-check, but for when we merge this, what kind of smoke tests should I make to make sure we haven't broken the user table?

bszabo commented:
Good question, @connorhaugh . I haven't actually changed the model (tables are all the same), so in theory if it builds and passes unit tests, it should be OK. But in theory, theory and practice are the same; in practice, they're not. I need to explore a bit to answer your question.

bszabo commented:
Practice rules, @connorhaugh ! The basic smoke test is to log in to LMS, go to the profile menu, and then save something like the maximum degree you attained. Then you log out, log back in, and see that what you selected is still there. 

Presently, works in the master branch. Doesn't work on the feature branch. 

I'll tag you here once I've got it working; there's nothing further for you to do with this PR until then.

Commit:
feat: TNL-10136 fix dependencies

Multiple commits squashed. Common theme was to resolve dependencies with master branch

Commit:
feat: TNL-10136 fix small rebase errors

Too many blank lines, missing import between student model files

bszabo commented:
@connorhaugh Once again ready for review. I rebased with master yesterday & got the code to where it passes CI checks and successfully runs on devstack. My suggested approach for testing is to run a few manual tests to verify proper end-to-end behavior with representative user model changes and representative course enrollment changes. I further propose a pair programming session where we visit the nature of the changes in the former `models.py` file (combined student and course enrollment models), and the new `user.py` and `course_enrollment.py` files. My contention is that the nature of the changes is such that either nothing will work or everything will work, and that the code review will back this up.

For the manual student model tests, I recommend logging in to your devstack LMS and making edits to your account profile. Log out, log back in, and confirm that the changes you made are visible.

For the manual course enrollment test, I recommend you enroll in a course in your localhost environment, log out, log back in, and confirm that you're still enrolled in that course.

Commit:
feat: TNL-10136 Duplicate lines


Remove block of 3 lines duplicated in last rebase exercise

connorhaugh reviewed (approved):
We've discussed this code and it looks good to me, we plan to release this with care.

bszabo commented:
@connorhaugh Merge to stage has begun
